### DISCUSSION TITLE
RFC: Query Suggestions

### DISCUSSION THREAD

**Discussion by Strift** [meilisearch/product#740] [Feedback & Feature Proposal]
🔗  Upvote this feature [in our roadmap](https://roadmap.meilisearch.com/c/31-query-suggestions)

# Summary

Introduce query suggestion indexes as part of [Meilisearch Cloud](https://meilisearch.com/cloud?utm_source=github). A query suggestion index is based on the [analytics](https://www.meilisearch.com/docs/learn/cloud/analytics?utm_source=github) from one or multiple indexes, referred to as source indexes.

# Proposal

In their Meilisearch Cloud dashboard, users can configure a **Query Suggestion index** and select one or several indexes as sources. The Query Suggestion Index will contain documents mapping to the most popular search queries for the selected indexes.

For each search query in the index, the following data points will be available:
- Number of searches (over the last 30 days)
- Click rate (over the last 30 days)
- Conversion rate (over the last 30 days)

In practice, documents in a Query Suggestion Index will look like this:

```json
{
	"query": "iphone",
	"nb_searches_with_results": 100,
	"click_rate": 0.5,
	"conversion_rate": 0.05
}
```

Users can use the [sorting API](https://www.meilisearch.com/docs/learn/fine_tuning_results/sorting?utm_source=github) to implement ranking by number of searches, click rate, or conversion rate.

# Motivation & Drawbacks

This first iteration of the query suggestions API exposes the raw data points from the search analytics. This leaves users flexible in which metric they choose for popularity ranking. However, we understand that this might be limiting for users who need more advanced formulas to compute popularity.

However, this proposal allows us to release a first iteration of query suggestions that can evolve in non-breaking ways.

The evolutions we are currently considering are:
- Allow users to update the query suggestion index to set manually, e.g. a `popularity` field based on a custom formula
- Allow further customization of the popularity formula via the Meilisearch Cloud dashboard

# Front-end implementation

We’re looking to provide a query suggestion API that integrates best with our users' front-end stack. In particular, we're interested in compatibility with [InstantSearch](https://github.com/algolia/instantsearch) and [Autocomplete](https://github.com/algolia/autocomplete).

Please fill out this [30-second form to let us know what JS library you’re using](https://forms.gle/DKD7x2W4PRP5RbEV9) 🙏

---

Thanks for reading! Let us know what you think 👇

**Comment by onurusluca** [comment]
I have been waiting for this for so long. This would be great!

btw form says: 

> You need permission
> This form can only be viewed by users in the owner's organisation.
> 
> Try contacting the owner of the form if you think that this is a mistake.

**Reply by Strift** [reply]
Sorry about this! 😓 

I fixed the permission, you should now have access to the survey: https://forms.gle/TZZhjZmc7kHwmrTq8

**Comment by urbantrout** [comment]
This sounds great! But is this also going to be available for self-hosted Meilisearch instances?

**Reply by Strift** [reply]
It's not planned for the release, mainly because it relies on our Cloud analytics stack (i.e., edge functions proxy). For now, the Meilisearch engine does not have built-in capabilities for that. 

However, I would love to write a guide explaining how to build a similar analytics and query suggestions stack for self-hosted users.

**Reply by urbantrout** [reply]
Thanks for the explanation!

And I would love to read more about building such thing for self-hosted instances!

**Reply by Fissium** [reply]
@Strift I'm looking forward to the guide!

**Comment by belwalshubham** [comment]
Is there currently no way to implement a suggestion using MeiliSearch? Could you please check the following issue?
https://stackoverflow.com/questions/78545320/how-can-i-implement-auto-suggestions-with-meilisearch-in-a-gatsby-and-strapi-pro

**Reply by gmourier** [reply]
Hey @belwalshubham, it depends on what you mean by suggestion. 

If you want to show terms by popularity that you are sure will bring results to your users, you can do it by:

- track the most popular searches over time your users clicked on results from
- create a query suggestion index where documents are composed of `query_terms` and `popularity`; the documents are made from what you tracked in step 1.
- apply a custom ranking rule to sort the documents of the query suggestion index in descending order (most popular will be returned first), `popularity:desc`
- implement a search bar that suggests those query terms by searching in the query suggestion index when the user types (the /multi-search endpoint is handy for that — if you want to search both the query suggestion index and, let's say, the product index where the actual documents are to display them as well)

We are currently working it on the Cloud to streamline everything and [may write a technical blog post](https://github.com/orgs/meilisearch/discussions/740#discussioncomment-9339625) to explain how we did it to the open-source community.

**Reply by belwalshubham** [reply]
Hey @gmourier, thanks for your suggestion but I don't want to display the term by popularity. I just want to show the closest suggestion for that term in the search bar.
